lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,267 @@
+	for _, s := range schedules {
+		responders = append(responders, Responder{
+			Type: "schedule",
+			ID:   s,
+		})
+	}
+	return responders
+}
+
+// PostReviewNote posts a note once a new request review appears.
+func (og OpsgenieClient) PostReviewNote(ctx context.Context, alertID string, review types.AccessReview) error {
+	note, err := og.buildReviewNoteBody(review)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/notes"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ResolveAlert resolves an alert and posts a note with resolution details.
+func (og OpsgenieClient) ResolveAlert(ctx context.Context, alertID string, resolution Resolution) error {
+	note, err := og.buildResolutionNoteBody(resolution)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/close"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (og OpsgenieClient) buildAlertBody(reqID string, reqData RequestData) (string, error) {
+	var requestLink string
+	if og.webProxyURL != nil {
+		reqURL := *og.webProxyURL
+		reqURL.Path = lib.BuildURLPath("web", "requests", reqID)
+		requestLink = reqURL.String()
+	}
+
+	var builder strings.Builder
+	err := alertBodyTemplate.Execute(&builder, struct {
+		ID          string
+		TimeFormat  string
+		RequestLink string
+		RequestData
+	}{
+		reqID,
+		time.RFC822,
+		requestLink,
+		reqData,
+	})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return builder.String(), nil
+}
+
+func (og OpsgenieClient) buildReviewNoteBody(review types.AccessReview) (string, error) {
+	var builder strings.Builder
+	err := reviewNoteTemplate.Execute(&builder, struct {
+		types.AccessReview
+		ProposedState string
+		TimeFormat    string
+	}{
+		review,
+		review.ProposedState.String(),
+		time.RFC822,
+	})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return builder.String(), nil
+}
+
+func (og OpsgenieClient) buildResolutionNoteBody(resolution Resolution) (string, error) {
+	var builder strings.Builder
+	err := resolutionNoteTemplate.Execute(&builder, struct {
+		Resolution    string
+		ResolveReason string
+	}{
+		string(resolution.Tag),
+		resolution.Reason,
+	})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return builder.String(), nil
+}
+
+// GetOnCall returns the list of responders on-call for a schedule.
+func (og OpsgenieClient) GetOnCall(ctx context.Context, scheduleName string) (RespondersResult, error) {
+	// v2/schedules/ScheduleName/on-calls?scheduleIdentifierType=name&flat=true'
```

Comment:
Explanations of query parameters (e.g. what is "flat"?) and a link to reference documentation would be useful.

Same for other methods that call Opsgenie APIs.